### PR TITLE
Get cherami and redis to be part of golang crossdock testing

### PR DESCRIPTION
**Pull request by peter-edge**
This PR sets up a new make target, `compose-codecov`, that will run all tests within a Docker container using docker-compose that is linked to running Cherami and Redis containers. Two environment variables, `CHERAMI=enabled` and `REDIS=enabled`, are also set on the test container, which now indicates to the tests in `internal/crossdock` to add `cherami,redis` as oneway transports to test. The net effect is that golang crossdock testing can now test the Cherami and Redis transports. The net effect for code coverage is to increase code coverage by about 1.5%. There is no effect on any other existing part of the setup, minus a few small changes to how code coverage is run, and some bug fixes.

I've also removed the split between experimental and main for Codecov - this PR brings the difference to a much smaller amount, and having both values displayed is a little confusing, and doesn't reflect the number we really track which is the overall code coverage percentage. The overall code coverage percentage is also what is displayed on the main Codecov page for yarpc-go and on the readme badge. I think that if code is worth being in yarpc-go, it should be equally tested.

One note is that the time that code coverage takes in Travis CI is significantly increased for a few reasons, however we can optimize this later.

I'll list specific changes so this is easier to follow:

- Update `.codecov.yml` to reflect these changes, and also make it a valid Codecov YAML file - you can check this with `curl --data-binary @.codecov.yml https://codecov.io/validate`.
- Up the target code coverage percentage to 80%.
- Update `.dockerignore` and `.gitignore` to only ignore `coverage.{html,txt}` instead of `coverage.{main,x}.{html,txt}`.
- Update `.travis.yml` to use this new setup, along with some cleanups.
- Update `Dockerfile.crossdock` per some cleanups in `etc/make`. I was also installing docker-compose in `Dockerfile.crossdock`, which didn't make sense.
- Use `healthcheck/redis` in docker-compose instead of `redis` https://github.com/docker-library/healthcheck/tree/master/redis.
- Use a `proxy_dep` in `docker-compose.yml` per https://github.com/docker/compose/issues/4369.
- Add a new `gotest` image in `docker-compose.yml` for testing. I can probably merge the go crossdock image with this later.
- Change `covermode` in `etc/bin/cover.sh` from `atomic` to `count`. `atomic` has a huge overhead, and doesn't really matter, it just makes sure your counts are thread-safe, and we don't really care about the counts (the counts of the number of times a given line is hit)
- Add a `CONCURRENCY` variable in `etc/bin/cover.sh` to easily adjust concurrency for coverage testing. Travis CI is having issues with Cherami when the tests are run concurrently, this can be fixed later.
- Fix an issue in `etc/bin/cover.sh` where if only a few packages were set, then `ignorePkgs` would be empty, resulting in an error when you piped to `grep -v "${ignorePkgs}"`.
- Move around some of the variables in the makefile files in `etc/make`, specifically relating to docker-compose.
- Add make targets `compose-test`, `compose-cover`, `compose-codecov`.
- Properly propagate Codecov environment variables to Docker when running code coverage for Codecov inside Docker.
- Update the `PACKAGES` variable in `etc/make/local.mk` to always have fully-qualified golang packages.  Code coverage needs the packages to be fully-qualified, so before this change, `PACKAGES=./internal/crossdock make cover` would fail, for example. The result of `go list ./internal/crossdock` is `go.uber.org/yarpc/internal/crossdock`.
- Add a conditional in `internal/crossdock/main_test.go` for `CHERAMI=enabled` and `REDIS=enabled` for oneway testing, similar to in `internal/crossdock/server/oneway/server.go`.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -11,30 +15,14 @@ cache:
...
         - secure: dhvshW7GvphJ6xvS8MxEtJGN+4C4d64hqOcaCKXXXVqyeunJ1gH/Cb6v1RcETgtRxOIHyyHo218j0cs0hNtIWCZq5YVbXMAJ5Sret+cKA3aZNMCbgwPNS48WdVZsCHurW22abq41ejfKuuFqSI+ooMprrsGeRjHMbJZM86k7UuEqxdvqGjotYugGcfiCjxnNtMHd1M6AQ9fPFzhg8TUeqPy7eDIdKTcqwa/h7ZsYn/WJp1LkUtwyHsfpweIv6JBQ9/xUDzjcoz5MA6Nj75/Mk4bJzzLEem3EKs1TTZOqV1NVt7VJ887NPe4fLIxP1/jQLzsh6kShTcctLszc70NgrowdwhdIqN/3TUmxkVbGP2JzbQ4yR6LVKCDDeWpKGiPsIKy75kEiW4ZlEZhnwCOKaR94FyBy1HEnhyoD9OJD6Gq/u4UdEw/QuNfE89gbr47LA4z487AzYGzH+1J3MlEbTNgTCmWduC0ghTx33vItrYMXXabGDe7XwfdJKTnMjfXVv+CVBdMAZVtDL27sEmbbAzIrRjP+JOFm7ZRYfzvxu5vCWlB/3wMMt7DzcutjxR4tyYs71Ec5zdUH/16wSvjQR1laJZCujlNSA/3FZUizCzW3qFVsBkaWRF+/C/cksi6cR83xIrc4y4WRLttAcbfqmsXaigL5Iz4GbLg62vmFP88=
-      sudo: required
-      dist: trusty
-      group: edge
-      services:
-        - docker
-    - env: CI_TYPES='deps examples codecov' SUPPRESS_DOCKER=1 SUPPRESS_CROSSDOCK=1
-      go: 1.7
```
Comment (kriskowal) [inline .travis.yml]: Do we intentionally drop 1.7 here?

File: docker-compose.yml [hunk 2]
```diff
@@ -123,10 +121,34 @@ services:
...
+            - redis
+            - cherami
+        build:
+          context: .
+          dockerfile: Dockerfile.1.8
+        environment:
+            - REDIS=enabled
+            - CHERAMI=enabled
+
+    #https://github.com/docker/compose/issues/4369
```
Comment (kriskowal) [inline docker-compose.yml]: Ah.

File: etc/bin/cover.sh [hunk 3]
```diff
@@ -20,6 +21,14 @@ mkdir -p "$COVER"
 
 ignorePkgs=""
 
+filterIgnorePkgs() {
+  if [[ -z "${ignorePkgs}" ]]; then
+    cat -
```
Comment (kriskowal) [inline etc/bin/cover.sh]: Superfluous `-`.

File: internal/crossdock/main_test.go [hunk 4]
```diff
@@ -168,7 +169,7 @@ func TestCrossdock(t *testing.T) {
 				"server_oneway": "127.0.0.1",
 			},
 			axes: axes{
-				"transport_oneway": []string{"http"},
+				"transport_oneway": getTransportOnewayAxes(),
```
Comment (kriskowal) [inline internal/crossdock/main_test.go]: Good enough for now. Might be worth clawing this back to be filtered through Crossdock environment variables.

File: .travis.yml (hunk 1 above)
Comment (peter-edge) [inline .travis.yml]: So this is something I want to revisit soon - I know we said we want to keep 1.7, and we can, it just means one more running test at this point, it's a pain in the butt to keep separate docker-compose images for each version.

Short: Yes, but I can re-add in a follow-up one-line PR if we want, it means more resources used in Travis CI though and I don't think it adds much value.

File: internal/crossdock/main_test.go (hunk 4 above)
Comment (peter-edge) [inline internal/crossdock/main_test.go]: I want to revisit all of this when we have time.

File: etc/bin/cover.sh (hunk 3 above)
Comment (peter-edge) [inline etc/bin/cover.sh]: Seems so!

**Pull request merged**